packages/vscode-extension/src/webview/views/PreviewView.tsx

```diff
@@ -126,6 +130,43 @@ function PreviewView() {
     }
   };
 
+  const MAX_RECORDING_TIME_SEC = 300;
+
+  const handleRecording = useCallback(async () => {
+    try {
+      if (!isRecording) {
+        setIsRecording(true);
+        project.startRecording();
+      } else {
+        const recordingData = await project.captureAndStopRecording();
+        if (recordingTime > 0) {
+          saveVideoRecording(recordingData);
+        }
+        setIsRecording(false);
+        setRecordingTime(0);
+      }
+    } catch (e) {
+      showDismissableError("Failed to capture recording");
+    }
+  }, [isRecording, recordingTime]);
+
+  useEffect(() => {
+    if (!isRecording) {
+      setRecordingTime(0);
+      return;
+    }
+
+    setTimeout(() => {
+      setRecordingTime((prevRecordingTime) => {
+        if (prevRecordingTime >= MAX_RECORDING_TIME_SEC - 1) {
+          handleRecording();
+          return MAX_RECORDING_TIME_SEC;
+        }
+        return prevRecordingTime + 1;
+      });
+    }, 1000);
```

Comment:
I don't understand this bit. Maybe it should use interval actually. Now it doesn't do effect cleanup which may result in this timer running indefinitely until the timeout is hit regardless of whether the video is stopped or not.

Reply:
I've tested it and it behaves weird. After stopping recording before the time limit, setTimeout runs once more with prevRecordingTime set to 0 by handleRecording.
Adding a cleanup function would be a better solution.